File: src/main/java/minic/uilocal/diagnostics/MiniCDiagnosticListFactory.java
package minic.uilocal;

import minic.uiapi.UiDiagnosticDto;
import minic.uiapi.UiGlobalDataDto;
import minic.uiapi.UiRealtimeAnalysisDto;
import minic.uiapi.UiSourceRangeDto;
import minic.uiapi.UiStageDataDto;

import java.util.List;

/**
 * 根据 UI API DTO 生成 diagnostics 列表。
 */
public final class MiniCDiagnosticListFactory {
    /**
     * 创建诊断项。
     *
     * @param stageData 当前阶段数据
     * @param globalData 全局数据
     * @return 诊断项
     */
    public List<MiniCDiagnosticItem> create(UiStageDataDto stageData, UiGlobalDataDto globalData) {
        return create(stageData, globalData, null);
    }

    /**
     * 创建诊断项。
     *
     * @param stageData 当前阶段数据
     * @param globalData 全局数据
     * @param realtimeAnalysis 实时分析数据
     * @return 诊断项
     */
    public List<MiniCDiagnosticItem> create(
            UiStageDataDto stageData,
            UiGlobalDataDto globalData,
            UiRealtimeAnalysisDto realtimeAnalysis
    ) {
        if (realtimeAnalysis != null && !realtimeAnalysis.diagnostics().isEmpty()) {
            return realtimeAnalysis.diagnostics().stream()
                    .map(diagnostic -> from(diagnostic, realtimeAnalysis))
                    .toList();
        }
        List<UiDiagnosticDto> diagnostics = globalData != null && !globalData.diagnostics().isEmpty()
                ? globalData.diagnostics()
                : stageData == null ? List.of() : stageData.diagnostics();
        return diagnostics.stream().map(this::from).toList();
    }

    private MiniCDiagnosticItem from(UiDiagnosticDto diagnostic) {
        return from(diagnostic, null);
    }

    private MiniCDiagnosticItem from(UiDiagnosticDto diagnostic, UiRealtimeAnalysisDto analysis) {
        SourceLocation location = analysis == null
                ? new SourceLocation(1, Math.max(1, diagnostic.startOffset() + 1))
                : locationAt(analysis.sourceText(), diagnostic.startOffset());
        return new MiniCDiagnosticItem(
                diagnostic.code(),
                diagnostic.severity(),
                diagnostic.message(),
                new UiSourceRangeDto(diagnostic.sourceName(), diagnostic.startOffset(), diagnostic.endOffset()),
                location.line(),
                location.column()
        );
    }

    private SourceLocation locationAt(String source, int offset) {
        int safeOffset = Math.max(0, Math.min(offset, source.length()));
        int line = 1;
        int column = 1;
        for (int index = 0; index < safeOffset; index++) {
            char value = source.charAt(index);
            if (value == '\n') {
                line++;
                column = 1;
            } else {
                column++;
            }
        }
        return new SourceLocation(line, column);
    }

    private record SourceLocation(int line, int column) {
    }
}

File: uiweb/src/diagnostics/MiniCDiagnosticListFactory.ts
import type { JavaMirrorFile } from "../translation/javaMirror";
import type {
  UiDiagnosticDto,
  UiGlobalDataDto,
  UiRealtimeAnalysisDto,
  UiStageDataDto,
} from "../translation/uiapi";
import { MiniCDiagnosticItem } from "./MiniCDiagnosticItem";

export const miniCDiagnosticListFactoryMirror = {
  "javaPath": "src/main/java/minic/uilocal/diagnostics/MiniCDiagnosticListFactory.java",
  "webPath": "uiweb/src/diagnostics/MiniCDiagnosticListFactory.ts",
  "packageName": "minic.uilocal",
  "exportName": "MiniCDiagnosticListFactory",
  "kind": "class",
  "imports": [
    "minic.uiapi.UiDiagnosticDto",
    "minic.uiapi.UiGlobalDataDto",
    "minic.uiapi.UiRealtimeAnalysisDto",
    "minic.uiapi.UiSourceRangeDto",
    "minic.uiapi.UiStageDataDto",
    "java.util.List"
  ],
  "fields": [],
  "methods": [
    {
      "name": "create",
      "signature": "create(UiStageDataDto stageData, UiGlobalDataDto globalData)"
    },
    {
      "name": "create",
      "signature": "create(UiStageDataDto stageData, UiGlobalDataDto globalData, UiRealtimeAnalysisDto realtimeAnalysis)"
    },
    {
      "name": "from",
      "signature": "from(UiDiagnosticDto diagnostic)"
    },
    {
      "name": "from",
      "signature": "from(UiDiagnosticDto diagnostic, UiRealtimeAnalysisDto analysis)"
    },
    {
      "name": "locationAt",
      "signature": "locationAt(String source, int offset)"
    },
    {
      "name": "SourceLocation",
      "signature": "SourceLocation(int line, int column)"
    }
  ]
} as const satisfies JavaMirrorFile;

export class MiniCDiagnosticListFactory {
  static readonly mirror = miniCDiagnosticListFactoryMirror;

  readonly mirror = miniCDiagnosticListFactoryMirror;

  create(
    stageData: UiStageDataDto | null,
    globalData: UiGlobalDataDto | null,
    realtimeAnalysis: UiRealtimeAnalysisDto | null = null,
  ): readonly MiniCDiagnosticItem[] {
    if (realtimeAnalysis !== null && realtimeAnalysis.diagnostics.length > 0) {
      return realtimeAnalysis.diagnostics.map((diagnostic) => this.from(diagnostic, realtimeAnalysis));
    }
    const diagnostics =
      globalData !== null && globalData.diagnostics.length > 0
        ? globalData.diagnostics
        : stageData === null
          ? []
          : stageData.diagnostics;
    return diagnostics.map((diagnostic) => this.from(diagnostic));
  }

  from(diagnostic: UiDiagnosticDto, analysis: UiRealtimeAnalysisDto | null = null): MiniCDiagnosticItem {
    const location =
      analysis === null
        ? { line: 1, column: Math.max(1, diagnostic.startOffset + 1) }
        : locationAt(analysis.sourceText, diagnostic.startOffset);
    return new MiniCDiagnosticItem(
      diagnostic.code,
      diagnostic.severity,
      diagnostic.message,
      {
        sourceName: diagnostic.sourceName,
        startOffset: diagnostic.startOffset,
        endOffset: diagnostic.endOffset,
      },
      location.line,
      location.column,
    );
  }

  summary(): string {
    return `MiniCDiagnosticListFactory: ${this.mirror.methods.length} methods, ${this.mirror.fields.length} fields`;
  }
}

interface SourceLocation {
  readonly line: number;
  readonly column: number;
}

function locationAt(source: string, offset: number): SourceLocation {
  const safeOffset = Math.max(0, Math.min(Math.trunc(offset), source.length));
  let line = 1;
  let column = 1;
  for (let index = 0; index < safeOffset; index += 1) {
    if (source[index] === "\n") {
      line += 1;
      column = 1;
    } else {
      column += 1;
    }
  }
  return { line, column };
}

export default MiniCDiagnosticListFactory;
